grammar.js

```diff
@@ -1115,29 +1123,113 @@ module.exports = grammar({
       optional(field('parameters', $.arguments))
     ),
 
-    string: $ => {
-      const b_prefix = /[bB]/
-      const single_quote_chars = repeat(/\\'|\\\\|\\?[^'\\]/)
-      const single_quote_string = seq(
-        optional(b_prefix), "'", single_quote_chars, "'"
+    _string_member_access_expression: $ => prec(PREC.MEMBER, seq(
+      field('object', $._string_dereferencable_expression),
+      '->',
+      $._member_name
+    )),
+
+    _string_dereferencable_expression: $ => prec(PREC.DEREF, choice(
+      $._variable_name,
+      alias($._string_subscript_expression, $.subscript_expression),
+      alias($._string_member_access_expression, $.member_access_expression)
+    )),
+
+    _string_subscript_expression: $ => seq(
+      $._string_dereferencable_expression,
+      choice(
+        seq('[', optional($._expression), ']'),
+        seq('{', $._expression, '}')
       )
+    ),
 
-      const double_quote_string = seq(
-        optional(b_prefix), '"', double_quote_chars(), '"'
+    _complex_string_part: $ => prec.right(seq(
+      "{",
+      choice(
+        alias($._string_member_access_expression, $.member_access_expression),
+        $._variable_name,
+        alias($._string_subscript_expression, $.subscript_expression),
+      ),
+      "}"
+    )),
+
+    _simple_string_member_access_expression: $ => seq(
+      field('object', $._simple_string_variable_name),
+      '->',
+      field('name', $.name),
+    ),
+
+    _simple_string_subscript_unary_expression: $ => prec.left(seq('-', $.integer)),
+
+    _simple_string_array_access_argument: $ => choice(
+      $.integer,
+      alias($._simple_string_subscript_unary_expression, $.unary_op_expression)
+    ),
+
+    _simple_string_subscript_expression: $ => seq(
+      $._simple_string_variable_name,
+      choice(
+        seq('[', $._simple_string_array_access_argument, ']'),
+        seq('{', $._simple_string_array_access_argument, '}')
       )
+    ),
 
-      return token(choice(
-        single_quote_string,
-        double_quote_string
-        // nowdoc_string,
-      ))
-    },
+    _simple_string_dynamic_variable_name: $ => seq('$', '{', '$', $.name, '}'),
+
+    _simple_string_variable_name: $ => choice(
+      alias($._simple_string_dynamic_variable_name, $.dynamic_variable_name),
+      $.variable_name
+    ),
+
+    _simple_string_part: $ => prec.right(choice(
```

Comment:
@maxbrunsfeld When you have the time (I fully understand if this is not a priority atm), could you give me a pointer on how to move forward with this?

For some reason, this rule always returns $._variable_name, but if I remove that option it wil correctly match the other rules.

I'm considering moving this part to the C++ based scanner where I'm able to fully understand and control what happens, but I'm simultaneously realizing that it's probably a matter of my tree-sitter incompetence

(yes the code needs a thorough clean up when everything is working, since the PHP strings are almost a separate grammar in itself)